extensions-core/druid-catalog/src/main/java/org/apache/druid/catalog/storage/sql/CatalogManager.java

```diff
@@ -0,0 +1,167 @@
+import org.apache.druid.catalog.model.TableMetadata;
+import org.apache.druid.catalog.model.TableSpec;
+
+import javax.annotation.Nullable;
+
+import java.util.List;
+import java.util.function.Function;
+
+/**
+ * Manages catalog data. Used in Coordinator, which will be in either
+ * an leader or standby state. The Coordinator calls the {@link #start()}
+ * method when it becomes the leader, and calls {@link #stop()} when
+ * it loses leadership, or shuts down.
+ *
+ * Performs detailed CRUD operations on the catalog tables table.
+ * Higher-level operations appear elsewhere.
+ */
+public interface CatalogManager
+{
+  /**
+   * Thrown with an "optimistic lock" fails: the version of a
+   * catalog object being updated is not the same as that of
+   * the expected version.
+   */
+  class OutOfDateException extends Exception
+  {
+    public OutOfDateException(String msg)
+    {
+      super(msg);
+    }
+  }
+
+  class NotFoundException extends Exception
+  {
+    public NotFoundException(String msg)
+    {
+      super(msg);
+    }
+  }
+
+  /**
+   * Indicates an attempt to insert a duplicate key into a table.
+   * This could indicate a logic error, or a race condition. It is
+   * generally not retryable: it us unrealistic to expect the other
+   * thread to helpfully delete the record it just added.
+   */
+  class DuplicateKeyException extends Exception
+  {
+    public DuplicateKeyException(String msg, Exception e)
+    {
+      super(msg, e);
+    }
+  }
+
+  interface Listener
+  {
+    void added(TableMetadata table);
+    void updated(TableMetadata table);
+    void deleted(TableId id);
+  }
+
+  void start();
+
+
+  void register(Listener listener);
+  void createTableDefnTable();
+
+  /**
+   * Create a table entry.
+   *
+   * @return the version of the newly created table. Call
+   * {@link TableMetadata#asUpdate(long)} if you want a new
+   * {@link TableMetadata} with the new version.
+   * @throws {@link DuplicateKeyException} if the row is a duplicate
+   * (schema, name) pair. This generally indicates a code error,
+   * or since our code is perfect, a race condition or a DB
+   * update outside of Druid. In any event, the error is not
+   * retryable: the user should pick another name, or update the
+   * existing table
+   */
+  long create(TableMetadata table) throws DuplicateKeyException;
+
+  /**
+   * Update a table definition.
+   * <p>
+   * If {@code oldVersion == 0}, overwrites any current content.
+   * This is a potential race conditions if this is a partial update
+   * because of the possibility of another user doing an update since the
+   * read. Fine when the goal is to replace the entire definition.
+   * Else, only does the update is at the given version.
+   * <p>
+   * Retryable only if the version is given, and an
+   * {@code OutOfDateException} is thrown.
+   */
+  long update(TableMetadata table, long oldVersion) throws OutOfDateException, NotFoundException;
```

Comment:
Should we have a separate version that does not take a version, so that this is always updating given a current version, and the other method could always just force update whatever the current version is?